runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/retries/delay/ClientRateLimiter.kt

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.retries.delay
+
+import aws.smithy.kotlin.runtime.InternalApi
+import aws.smithy.kotlin.runtime.retries.policy.RetryErrorType
+
+/**
+ * A client-side utility that can limit the rate of transactions. This limiter is usually backed by some set of permits
+ * which refill over time. This kind of rate limiting is often used in conjunction with a retry algorithm, especially
+ * when it may be desirable to delay the initial attempt of a transaction (instead of just delaying subsequent
+ * attempts).
+ *
+ * Callers should invoke [acquire] with the cost of their transaction prior to executing. If
+ * there are not enough permits available currently, the call may delay.
+ *
+ * Once the transaction is completed, callers should invoke [update] and indicate the type of error (if any) the
+ * transaction yielded. The rate limiter may use this information to adjust the number of available permits or the rate
+ * of permit renewal.
+ */
+public interface ClientRateLimiter {
+    public val config: Config
+
+    /**
+     * Acquire a "permit" to conduct a transaction. If not enough permits are available, this method may delay.
+     * @param cost The relative cost of the transaction. Some transactions may take more work or represent a bundle of
+     * smaller transactions.
+     */
+    public suspend fun acquire(cost: Int)
```

Comment:
**comment**: It seems odd to have an `acquire` but no `release` sibling. I'm assuming `update` is that but it feels sort of clunky (to me anyway) to not have the API symmetry.

Reply:
Generally, rate limiters don't allow "releasing" capacity. You must acquire a permit but the only thing that reincrements permits is time, not successes or failures. (A _concurrency_ limiter would have symmetrical acquire and release methods.) See for example [Guava's `RateLimiter`](https://guava.dev/releases/19.0/api/docs/index.html?com/google/common/util/concurrent/RateLimiter.html).

The `update` method is not analogous to releasing capacity. It's a mechanism for dynamically adjusting the refill rate based on the type of response received from the last piece of work.